src/vs/platform/remote/node/sharedProcessTunnelService.ts

```diff
@@ -0,0 +1,80 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ILogService } from 'vs/platform/log/common/log';
+import { ISharedProcessTunnel, ISharedProcessTunnelService } from 'vs/platform/remote/common/sharedProcessTunnelService';
+import { ITunnelService, RemoteTunnel } from 'vs/platform/remote/common/tunnel';
+import { IAddress, IAddressProvider } from 'vs/platform/remote/common/remoteAgentConnection';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { canceled } from 'vs/base/common/errors';
+
+export class SharedProcessTunnelService extends Disposable implements ISharedProcessTunnelService {
+	_serviceBrand: undefined;
+
+	private static _lastId = 0;
+
+	private readonly _tunnels: Map<string, RemoteTunnel> = new Map<string, RemoteTunnel>();
+	private readonly _disposedTunnels: Set<string> = new Set<string>();
+
+	constructor(
+		@ITunnelService private readonly _tunnelService: ITunnelService,
+		@ILogService private readonly _logService: ILogService,
+	) {
+		super();
+	}
+
+	public override dispose(): void {
+		super.dispose();
+		this._tunnels.forEach((tunnel) => tunnel.dispose());
+	}
+
+	async createTunnel(): Promise<{ id: string }> {
+		const id = String(++SharedProcessTunnelService._lastId);
+		return { id };
+	}
+
+	async startTunnel(id: string, address: IAddress, tunnelRemoteHost: string, tunnelRemotePort: number, tunnelLocalPort: number | undefined, elevateIfNeeded: boolean | undefined): Promise<ISharedProcessTunnel> {
+		const addressProvider = new class implements IAddressProvider {
+			async getAddress(): Promise<IAddress> {
+				return address;
+			}
+		};
+
+		const tunnel = await Promise.resolve(this._tunnelService.openTunnel(addressProvider, tunnelRemoteHost, tunnelRemotePort, tunnelLocalPort, elevateIfNeeded));
```

Comment:
Why `await Promise.resolve`?

Reply:
Because `ITunnelService.openTunnel` returns `Promise<RemoteTunnel | undefined> | undefined`, and `Promise.resolve(Promise<RemoteTunnel | undefined> | undefined)` simplifies that a bit to `Promise<RemoteTunnel | undefined>`